File: tests/cpp/core/schema/test_head.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated HeadPose FlatBuffer message.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>

// Include generated FlatBuffer headers.
#include <schema/head_generated.h>
#include <schema/timestamp_generated.h>

#include <type_traits>

// =============================================================================
// Compile-time verification of FlatBuffer field IDs.
// These ensure schema field IDs remain stable across changes.
// VT values are computed as: (field_id + 2) * 2.
// =============================================================================
#define VT(field) (field + 2) * 2
static_assert(core::HeadPose::VT_POSE == VT(0));
static_assert(core::HeadPose::VT_IS_VALID == VT(1));
static_assert(core::HeadPose::VT_IS_TRACKED == VT(2));

static_assert(core::HeadPoseRecord::VT_DATA == VT(0));

// =============================================================================
// Compile-time verification of FlatBuffer field types.
// These ensure schema field types remain stable across changes.
// =============================================================================
#define TYPE(field) decltype(std::declval<core::HeadPose>().field())
static_assert(std::is_same_v<TYPE(pose), const core::Pose*>);
static_assert(std::is_same_v<TYPE(is_valid), bool>);
static_assert(std::is_same_v<TYPE(is_tracked), bool>);


TEST_CASE("HeadPoseT can handle is_valid value properly", "[head][native]")
{
    flatbuffers::FlatBufferBuilder builder(1024);

    // Create HeadPoseT with default values, is_valid should be false.
    auto head_pose = std::make_unique<core::HeadPoseT>();
    CHECK(head_pose->is_valid == false);
    CHECK(head_pose->is_tracked == false);

    // Set is_valid to true.
    head_pose->is_valid = true;
    CHECK(head_pose->is_valid == true);
}

TEST_CASE("HeadPoseT can handle is_tracked value properly", "[head][native]")
{
    auto head_pose = std::make_unique<core::HeadPoseT>();
    CHECK(head_pose->is_tracked == false);

    head_pose->is_tracked = true;
    CHECK(head_pose->is_tracked == true);
}

TEST_CASE("HeadPoseT default construction", "[head][native]")
{
    auto head_pose = std::make_unique<core::HeadPoseT>();

    // Default values.
    CHECK(head_pose->pose == nullptr);
    CHECK(head_pose->is_valid == false);
    CHECK(head_pose->is_tracked == false);
}

TEST_CASE("HeadPoseT can store pose data", "[head][native]")
{
    auto head_pose = std::make_unique<core::HeadPoseT>();

    // Create and set pose.
    core::Point position(1.5f, 2.5f, 3.5f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f); // Identity quaternion.
    head_pose->pose = std::make_unique<core::Pose>(position, orientation);

    CHECK(head_pose->pose->position().x() == Catch::Approx(1.5f));
    CHECK(head_pose->pose->position().y() == Catch::Approx(2.5f));
    CHECK(head_pose->pose->position().z() == Catch::Approx(3.5f));
    CHECK(head_pose->pose->orientation().x() == Catch::Approx(0.0f));
    CHECK(head_pose->pose->orientation().y() == Catch::Approx(0.0f));
    CHECK(head_pose->pose->orientation().z() == Catch::Approx(0.0f));
    CHECK(head_pose->pose->orientation().w() == Catch::Approx(1.0f));
}

TEST_CASE("HeadPoseT can store rotation quaternion", "[head][native]")
{
    auto head_pose = std::make_unique<core::HeadPoseT>();

    // 90-degree rotation around Z axis.
    core::Point position(0.0f, 0.0f, 0.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.7071068f, 0.7071068f); // (x, y, z, w)
    head_pose->pose = std::make_unique<core::Pose>(position, orientation);

    CHECK(head_pose->pose->orientation().z() == Catch::Approx(0.7071068f).epsilon(0.0001));
    CHECK(head_pose->pose->orientation().w() == Catch::Approx(0.7071068f).epsilon(0.0001));
}

TEST_CASE("HeadPoseT serialization and deserialization", "[head][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(1024);

    // Create HeadPoseT with all fields set.
    auto head_pose = std::make_unique<core::HeadPoseT>();
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    head_pose->pose = std::make_unique<core::Pose>(position, orientation);
    head_pose->is_valid = true;
    head_pose->is_tracked = true;

    // Serialize.
    auto offset = core::HeadPose::Pack(builder, head_pose.get());
    builder.Finish(offset);

    // Deserialize.
    auto buffer = builder.GetBufferPointer();
    auto deserialized = flatbuffers::GetRoot<core::HeadPose>(buffer);

    // Verify.
    CHECK(deserialized->pose()->position().x() == Catch::Approx(1.0f));
    CHECK(deserialized->pose()->position().y() == Catch::Approx(2.0f));
    CHECK(deserialized->pose()->position().z() == Catch::Approx(3.0f));
    CHECK(deserialized->is_valid() == true);
    CHECK(deserialized->is_tracked() == true);
}

TEST_CASE("HeadPoseT can be unpacked from buffer", "[head][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(1024);

    // Create and serialize.
    auto original = std::make_unique<core::HeadPoseT>();
    core::Point position(5.0f, 6.0f, 7.0f);
    core::Quaternion orientation(0.1f, 0.2f, 0.3f, 0.9f);
    original->pose = std::make_unique<core::Pose>(position, orientation);
    original->is_valid = true;
    original->is_tracked = true;

    auto offset = core::HeadPose::Pack(builder, original.get());
    builder.Finish(offset);

    // Unpack to HeadPoseT.
    auto buffer = builder.GetBufferPointer();
    auto head_pose_fb = flatbuffers::GetRoot<core::HeadPose>(buffer);
    auto unpacked = std::make_unique<core::HeadPoseT>();
    head_pose_fb->UnPackTo(unpacked.get());

    CHECK(unpacked->pose->position().x() == Catch::Approx(5.0f));
    CHECK(unpacked->pose->position().y() == Catch::Approx(6.0f));
    CHECK(unpacked->pose->position().z() == Catch::Approx(7.0f));
    CHECK(unpacked->is_valid == true);
    CHECK(unpacked->is_tracked == true);
}

// =============================================================================
// HeadPoseRecord Tests (timestamp lives on the Record wrapper)
// =============================================================================
TEST_CASE("HeadPoseRecord serialization with DeviceDataTimestamp", "[head][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(1024);

    auto record = std::make_shared<core::HeadPoseRecordT>();
    record->data = std::make_shared<core::HeadPoseT>();
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    record->data->pose = std::make_unique<core::Pose>(position, orientation);
    record->data->is_valid = true;
    record->data->is_tracked = true;
    record->timestamp = std::make_shared<core::DeviceDataTimestamp>(1000000000LL, 2000000000LL, 3000000000LL);

    auto offset = core::HeadPoseRecord::Pack(builder, record.get());
    builder.Finish(offset);

    auto deserialized = flatbuffers::GetRoot<core::HeadPoseRecord>(builder.GetBufferPointer());

    CHECK(deserialized->timestamp()->available_time_local_common_clock() == 1000000000LL);
    CHECK(deserialized->timestamp()->sample_time_local_common_clock() == 2000000000LL);
    CHECK(deserialized->timestamp()->sample_time_raw_device_clock() == 3000000000LL);
    CHECK(deserialized->data()->pose()->position().x() == Catch::Approx(1.0f));
    CHECK(deserialized->data()->is_valid() == true);
    CHECK(deserialized->data()->is_tracked() == true);
}

TEST_CASE("HeadPoseRecord can be unpacked with DeviceDataTimestamp", "[head][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(1024);

    auto original = std::make_shared<core::HeadPoseRecordT>();
    original->data = std::make_shared<core::HeadPoseT>();
    original->data->is_valid = true;
    original->data->is_tracked = false;
    original->timestamp = std::make_shared<core::DeviceDataTimestamp>(111LL, 222LL, 333LL);

    auto offset = core::HeadPoseRecord::Pack(builder, original.get());
    builder.Finish(offset);

    auto fb = flatbuffers::GetRoot<core::HeadPoseRecord>(builder.GetBufferPointer());
    auto unpacked = std::make_shared<core::HeadPoseRecordT>();
    fb->UnPackTo(unpacked.get());

    CHECK(unpacked->timestamp->available_time_local_common_clock() == 111LL);
    CHECK(unpacked->timestamp->sample_time_local_common_clock() == 222LL);
    CHECK(unpacked->timestamp->sample_time_raw_device_clock() == 333LL);
    CHECK(unpacked->data->is_valid == true);
    CHECK(unpacked->data->is_tracked == false);
}

File: tests/python/core/schema/test_head.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for HeadPose in isaacteleop.schema.

HeadPose is a FlatBuffers table that represents head pose data:
- pose: The Pose struct (position and orientation)
- is_valid: Whether the pose value is valid to read (OpenXR VALID)
- is_tracked: Whether the pose is actively tracked (OpenXR TRACKED)

Timestamps are carried by HeadPoseRecord, not HeadPose.

Note: Python code should only READ this data (created by C++ trackers), not modify it.
"""

import pytest

from isaacteleop.schema import (
    HeadPose,
    HeadPoseRecord,
    Pose,
    Point,
    Quaternion,
    DeviceDataTimestamp,
)


class TestHeadPoseTConstruction:
    """Tests for HeadPose construction and basic properties."""

    def test_default_construction(self):
        """Test default construction creates HeadPose with default-initialized fields."""
        head_pose = HeadPose()

        assert head_pose is not None
        assert head_pose.pose is not None
        assert head_pose.is_valid is False
        assert head_pose.is_tracked is False

    def test_parameterized_construction(self):
        """Test construction with pose, is_valid, and is_tracked."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        head_pose = HeadPose(pose, True, True)

        assert head_pose.pose.position.x == pytest.approx(1.0)
        assert head_pose.pose.position.y == pytest.approx(2.0)
        assert head_pose.pose.position.z == pytest.approx(3.0)
        assert head_pose.pose.orientation.w == pytest.approx(1.0)
        assert head_pose.is_valid is True
        assert head_pose.is_tracked is True

    def test_parameterized_construction_defaults_is_tracked_false(self):
        """Two-arg constructor keeps is_tracked=False for back-compat."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        head_pose = HeadPose(pose, True)

        assert head_pose.is_valid is True
        assert head_pose.is_tracked is False


class TestHeadPoseTRepr:
    """Tests for HeadPose __repr__ method."""

    def test_repr_default(self):
        """Test __repr__ with default construction."""
        head_pose = HeadPose()

        repr_str = repr(head_pose)
        assert "HeadPose" in repr_str

    def test_repr_with_values(self):
        """Test __repr__ with parameterized construction."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        head_pose = HeadPose(pose, True, True)

        repr_str = repr(head_pose)
        assert "HeadPose" in repr_str
        assert "is_valid=True" in repr_str
        assert "is_tracked=True" in repr_str


class TestHeadPoseRecordTimestamp:
    """Tests for HeadPoseRecord with DeviceDataTimestamp."""

    def test_construction_with_timestamp(self):
        """Test HeadPoseRecord carries DeviceDataTimestamp."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        data = HeadPose(pose, True, True)
        ts = DeviceDataTimestamp(1000000000, 2000000000, 3000000000)
        record = HeadPoseRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 1000000000
        assert record.timestamp.sample_time_local_common_clock == 2000000000
        assert record.timestamp.sample_time_raw_device_clock == 3000000000
        assert record.data.is_valid is True
        assert record.data.is_tracked is True

    def test_payload_less_record(self):
        """A record may carry a timestamp and no payload: MCAP's frame sentinel."""
        record = HeadPoseRecord(None, DeviceDataTimestamp(1, 2, 3))
        assert record.data is None
        assert record.timestamp.available_time_local_common_clock == 1

    def test_timestamp_fields(self):
        """Test all three DeviceDataTimestamp fields are accessible."""
        data = HeadPose()
        ts = DeviceDataTimestamp(111, 222, 333)
        record = HeadPoseRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 111
        assert record.timestamp.sample_time_local_common_clock == 222
        assert record.timestamp.sample_time_raw_device_clock == 333
